airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py

```diff
@@ -0,0 +1,260 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, Iterable, Mapping, Optional
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.declarative.stream_slicers.stream_slicer import StreamSlicer
+from airbyte_cdk.sources.declarative.types import Record, StreamSlice, StreamState
+
+
+class Hashabledict(dict):
+    """
+    We are concerned of the performance of looping through the `states` list and evaluating equality on the partition. To reduce this
+    concern, we wanted to use dictionaries to map `partition -> cursor`. However, partitions are dict and dict can't be used as dict keys
+    since they are not hashable. By creating hashable key and values, we can implement __hash__ for dict and then use them as dict keys.
+    Note that those dicts shouldn't be modified has their hash would change. For our case, it should be fine as partitions can be seen as
+    immutable.
+    """
+
+    def __hash__(self):
+        return hash(self._to_hashable(self))
+
+    def _to_hashable(self, value):
+        if type(value) == list:
+            return (self._to_hashable(item) for item in value)
+        elif type(value) in {dict, Hashabledict}:
+            return frozenset(value), frozenset((self._to_hashable(item) for item in value.values()))
+        return value
+
+
+class PerPartitionStreamSlice(StreamSlice):
+    def __init__(self, partition: Mapping[str, Any], cursor_slice: Mapping[str, Any]):
+        self._partition = partition
+        self._cursor_slice = cursor_slice
+        self._stream_slice = dict(partition) | dict(cursor_slice)
+
+    @property
+    def partition(self):
+        return self._partition
+
+    @property
+    def cursor_slice(self):
+        return self._cursor_slice
+
+    def __repr__(self):
+        return repr(self._stream_slice)
+
+    def __setitem__(self, key: str, value: Any):
+        raise ValueError("PerPartitionStreamSlice is immutable")
+
+    def __getitem__(self, key: str):
+        return self._stream_slice[key]
+
+    def __len__(self):
+        return len(self._stream_slice)
+
+    def __iter__(self):
+        return iter(self._stream_slice)
+
+    def __contains__(self, item: str):
+        return item in self._stream_slice
+
+    def keys(self):
+        return self._stream_slice.keys()
+
+    def items(self):
+        return self._stream_slice.items()
+
+    def values(self):
+        return self._stream_slice.values()
+
+    def get(self, key: str, default: Any) -> Any:
+        return self._stream_slice.get(key, default)
+
+    def __eq__(self, other):
+        if type(other) == dict:
+            return self._stream_slice == other
+        if type(other) == PerPartitionStreamSlice:
+            # noinspection PyProtectedMember
+            return self._partition == other._partition and self._cursor_slice == other._cursor_slice
+        return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class CursorFactory:
```

Comment:
Why is this needed?

Reply:
On my first iteration, I had a case where I wanted the `CursorFactory` to provide the cursor field. I later found a better solution and scraped that. However, it felt like just passing a `Callable[[], StreamSlicer]` might incur some refactoring if we see the need to add some behavior on `CursorFactory`. Doest that make sense?